File: C++/Maths/combinations.cpp
//COMBINATION LIBRARY

//Calculation C(n, r) modulo any number (prime or composite) using pascal triangle in O(n^2) complexity as precomputation and O(1) for retrevial
const int MAX = 1e3 + 3;
const int MOD = 1e9 + 7;

int ncr[MAX][MAX];

void pascal() {
	for (int i=0; i<MAX; ++i) {
		for (int j=0; j<=i; ++j) {
			if (j==0 || j==i) ncr[i][j] = 1;
			else ncr[i][j] = ncr[i-1][j-1] + ncr[i-1][j];
			if (ncr[i][j]>=MOD) ncr[i][j] -= MOD;
		}
	}
}

//space optimised version of above function
//dp state depends on only previous row elements so space can be reduced by factor of n
//for C(n, r) use ncr[n&1][r]
const int MAX = 1e3 + 3;
const int MOD = 1e9 + 7;

int ncr[2][MAX];

void pascal() {
	for (int i=0; i<MAX; ++i) {
		for (int j=0; j<=i; ++j) {
			if (j==0 || j==i) ncr[i&1][j] = 1;
			else ncr[i&1][j] = ncr[(i-1)&1][j-1] + ncr[(i-1)&1][j];
			if (ncr[i&1][j]>=MOD) ncr[i&1][j] -= MOD;
		}
	}
}

//Calculation of C(n, r) modulo prime in O(1) complexity
typedef long long LL;
const int MAX = 1e5 + 5;
const int MOD = 1e9 + 7;		//a prime number

int add(int a, int b, int c){int res=a+b;return(res>=c?res-c:res);}
int mod_neg(int a, int b, int c){int res;if(abs(a-b)<c)res=a-b;else res=(a-b)%c;return(res<0?res+c:res);}
int mul(int a, int b, int c){LL res=(LL)a*b;return(res>=c?res%c:res);}
template<typename T>T power(T e, T n, T m){T x=1,p=e;while(n){if(n&1)x=mul(x,p,m);p=mul(p,p,m);n>>=1;}return x;}
template<typename T>T extended_euclid(T a, T b, T &x, T &y){T xx=0,yy=1;y=0;x=1;while(b){T q=a/b,t=b;b=a%b;a=t;\
t=xx;xx=x-q*xx;x=t;t=yy;yy=y-q*yy;y=t;}return a;}
template<typename T>T mod_inverse(T a, T n){T x,y,z=0;T d=extended_euclid(a,n,x,y);return(d>1?-1:mod_neg(x,z,n));}

int fact[MAX], invp[MAX];

//Complexity : O(nlog n)
void init() {
	fact[0] = invp[0] = 1;
	for(int i=1; i<MAX; ++i) {
		fact[i] = mul(fact[i-1], i, MOD);
		invp[i] = mod_inverse(fact[i], MOD);
	}
}

//Complexity : O(1)
int ncr(int n, int r) {
	if (r<0 || r>n) return 0;
	return mul(fact[n], mul(invp[n-r], invp[r], MOD), MOD);
}

//Calculation of C(n, r) modulo any number(prime or composite) in O(n logn)
//Method is to find the power of all primes in n! and multiply pow(prime, count) modulo num to the result
//Sieve stores smallest prime factor for every number 
//for prime it stores number itself, so checking of number to be prime or not can be done in O(1)
const int MAX = 1e5 + 5;
const int MOD = 100;		// can be any number
int add(int a, int b, int c){int res=a+b;return(res>=c?res-c:res);}
int mod_neg(int a, int b, int c){int res;if(abs(a-b)<c)res=a-b;else res=(a-b)%c;return(res<0?res+c:res);}
int mul(int a, int b, int c){LL res=(LL)a*b;return(res>=c?res%c:res);}
template<typename T>T power(T e, T n, T m){T x=1,p=e;while(n){if(n&1)x=mul(x,p,m);p=mul(p,p,m);n>>=1;}return x;}

int lp[MAX];
vector<int> primes;

//Complexity: O(n) 
void factor_sieve() {
	lp[1] = 1;
	for (int i=2; i<MAX; ++i) {
		if (lp[i] == 0) {
			lp[i] = i;
			primes.emplace_back (i);
		}
		for (int j=0; j<primes.size() && primes[j]<=lp[i]; ++j) {
			int x = i * primes[j];
			if (x >= MAX) break;
			lp[x] = primes[j];
		}
	}
}

//calculates maximum value of i such that p^i divided n!
//p should be prime
int count_fact(int n, int p) {
	int k = 0; 
	while (n) {
		k += n/p;
		n /= p;
	}
	return k;
}
 
int ncr(int n, int r) {
	if (n < r) return 0;
	if (n==r || r==0) return 1;
	int ans = 1;
	for(int i=2; i<=n; ++i) {
		if (lp[i]==i) {		// if number is prime
			int cnt = count_fact(n, i) - count_fact(n-r, i) - count_fact(r, i);
			ans = mul(ans, power(i, cnt, MOD), MOD);
		}
	}
	return ans;
}

File: Python3/Maths/combinations.py
#General library for finding nCr modulo m
#where m, n, r can be any integer without any limits
#For usge just call the function ncr(n, r, m)

def factorise(m):
	ans = []
	p = 2
	while(m != 1):
		if (m%p != 0):
			p += 1
			continue
		e = 0
		while(m % p == 0) :
			m //= p
			e += 1
		ans.append([p, e])
		p += 1
	return ans

def mulmod(a, b, c):
	res = a * b
	if (res >= c):
		res %= c
	return res

def mod_inverse(a, n):
	N = n
	xx = 0
	yy = 1
	y = 0
	x = 1
	while(n > 0):
		q = a // n
		t = n
		n = a % n
		a = t
		t = xx
		xx = x - q * xx
		x = t
		t = yy
		yy = y - q * yy
		y = t
	x %= N
	x += N
	x %= N
	return x

def expo(a, b):
	x = 1
	y = a
	while(b > 0):
		if (b&1):
			x = x * y
		y = y * y
		b >>= 1
	return x

def power(a, b, c):
	x = 1
	y = a
	while(b > 0):
		if (b&1):
			x = mulmod(x, y, c)
		y = mulmod(y, y, c)
		b >>= 1
	return x

def init(p, pk):
	fact = []
	fact.append(1)
	for i in range(1, pk):
		red = i
		if (red % p == 0):
			red = 1
		fact.append(mulmod(fact[i-1], red, pk))
	return fact

def fact_mod(n, p, pk, fact):
	res = 1
	while(n > 0):
		times = n//pk
		res = mulmod(res, power(fact[pk-1], times, pk), pk)
		res = mulmod(res, fact[n%pk], pk)
		n //= p
	return res

def count_fact(n, p):
	ans = 0
	while(n > 0):
		ans += n//p
		n //= p
	return ans

def ncr_pk(n, r, p, e):
	if (r > n or r < 0):
		return 0
	if (r == 0 or n == r):
		return 1
	_e = count_fact(n, p) - count_fact(r, p) - count_fact(n-r, p)
	assert(_e >= 0)
	if (_e >= e):
		return 0
	pk = expo(p, e)
	fact = init(p, pk)
	ans = fact_mod(n, p, pk, fact)
	ans = mulmod(ans, mod_inverse(fact_mod(r, p, pk, fact), pk), pk)
	ans = mulmod(ans, mod_inverse(fact_mod(n-r, p, pk, fact), pk), pk)
	ans = mulmod(ans, expo(p, _e), pk)
	return ans

def pre_process(rem, mods):
	crt = []
	a = 1
	b = 1
	m = mods[0]
	crt.append([mods[0], a, b])
	L =	len(mods)
	for i in range(1, L):
		a = mod_inverse(m, mods[i])
		b = mod_inverse(mods[i], m)
		crt.append([m, a, b])
		m *= mods[i]
	assert(len(crt) == len(mods))
	return crt

def find_crt(rem, mods, crt):
	assert(len(crt) == len(mods))
	ans = rem[0]
	m = crt[0][0]
	L = len(mods)
	for i in range(1, L):
		a = crt[i][1]
		b = crt[i][2]
		m = crt[i][0]
		_m = m * mods[i]
		ans = mulmod(ans, b * mods[i], _m)
		ans = (ans + mulmod(rem[i], a * m, _m)) % (_m)
	return ans

def ncr(n, r, m):
	pf = factorise(m)
	rem = []
	mods = []
	L = len(pf)
	for i in range(L):
		get = ncr_pk(n, r, pf[i][0], pf[i][1])
		rem.append(get)
		mods.append(expo(pf[i][0], pf[i][1]))
	crt = pre_process(rem, mods)
	ans = find_crt(rem, mods, crt)
	return ans
